### PR TITLE
Expanded math stdlib

### PR DESCRIPTION
**Pull request by AntimatterReactor**
Unlike darkdarcool's random function, added a randrange function that actually works.
Also added cos and tan. Also also, added dirty parse function impl to
Types so it's shorter and more consice to convert to primitives.

### CODE REVIEW

**Review by darkdarcool — APPROVED** [review]
LGTM

File: Cargo.toml [hunk 1]
```diff
@@ -8,3 +8,4 @@ description = "A fun language with no classes whatsoever"
 clap = "^2.34"
 maplit = "^1.0"
 rustyline = "^9.1"
+rand = "^0.8"
```
Comment (19wintersp) [inline Cargo.toml:11]: Please can you put these in alphabetical order? Sorry it's just annoying me a tiny bit lol

File: src/types/tostring.rs [hunk 2]
```diff
@@ -34,6 +35,9 @@ impl Type {
             }
         }
     }
+    pub fn parse<F: FromStr>(&self) -> Result<F, F::Err> {
```
Comment (19wintersp) [inline src/types/tostring.rs:38]: What's the point of this function? Casting? If so, I think it'd be more efficient to try on a case-by-case basis.

File: src/stdlib/math.rs [hunk 3]
```diff
@@ -18,9 +20,30 @@ pub fn new() -> Module {
...
+
+            // RNG functions
+            "randrange".into() => native_func!(|_, args| {
+                Ok(Type::Float(thread_rng().gen_range(
+                    args[0].parse::<i32>()
+                        .expect("Invalid Argument: Type is not Int")..
+                    args[1].parse::<i32>()
+                        .expect("Invalid Argument: Type is not Int")
+                ) as f32))
+            }, 2)
```
Comment (19wintersp) [inline src/stdlib/math.rs:46]: I don't like how these will panic when the argument is wrong. These kinds of errors should be handled within the language, and catchable. Could you instead try to ensure the argument is of `Type::Float`?

File: src/stdlib/math.rs [hunk 4]
```diff
@@ -18,9 +20,30 @@ pub fn new() -> Module {
...
+                    .expect("Invalid Argument: Type is not Float").cos()))
+            }, 1),
+            "tan".into() => native_func!(|_, args| {
+                Ok(Type::Float(args[0].parse::<f32>()
+                    .expect("Invalid Argument: Type is not Float").tan()))
+            }, 1),
+
+            // RNG functions
+            "randrange".into() => native_func!(|_, args| {
+                Ok(Type::Float(thread_rng().gen_range(
```
Comment (19wintersp) [inline src/stdlib/math.rs:40]: Ideally, `thread_rng` would be called once per *thread*, as per its design, instead of instantiating it every time it is required. If this isn't possible (and it probably isn't with the current module system), using the standard `rand::random` function may be more idiomatic.

**Review by 19wintersp — CHANGES_REQUESTED** [review]
Thanks for contributing! I've left a few comments for review.

Overall I think it's somewhat blighted by the current system Europa uses for modules and native functions. (CC @cursorweb for this)

**Comment by cursorweb** [comment]
A couple of things I have noticed that the lang will need to have in order for these functions to be truly idiomatic
- [ ] Type checking
- [ ] Arity checking
- [ ] Type conversions
- [ ] Function errors
- [ ] Better module making 
at the moment I will therefore have to deny this pull request until these features have been implemented

File: src/types/tostring.rs (hunk 2 above)
Comment (AntimatterReactor) [reply src/types/tostring.rs:38]: I've tried doing it case by case, but the problem is that primitive types just sorta don't have an easy way to
cast it to other primitives. For example to cast f32 to int we can use `<F: From<f32>>` and `F::from(n)`. But if the F is an f32,
this wouldn't work. Any ideas?

File: Cargo.toml (hunk 1 above)
Comment (AntimatterReactor) [reply Cargo.toml:11]: Ok, will do

File: src/types/tostring.rs (hunk 2 above)
Comment (19wintersp) [reply src/types/tostring.rs:38]: I know how unhelpful it is, but my instinct is to say "do it any way other than this!"

I don't think it's idiomatic of Europa nor very good as a design decision in general to allow random types to be interpreted as others simply by nature of their stringified form (intended for **debugging**) being parsable to the type. In the situations in your code where you have used this function, there is a much better way of doing it, and on which I've left more commentary in the appropriate review comment. In short, you should match the type structure instead of casting through parsing debug strings.

File: src/stdlib/math.rs (hunk 3 above)
Comment (AntimatterReactor) [reply src/stdlib/math.rs:46]: Ah, I see. I just haven't analyzed the whole language interpreter, so I haven't realized that there is a way to raise error without just making the program panic. Definitely will change those to TypeError-throws

File: src/stdlib/math.rs (hunk 4 above)
Comment (AntimatterReactor) [reply src/stdlib/math.rs:40]: Oh, yeah. I haven't really delve too deep into the rand library, so I only have surface-level knowledge about it. Also will do.

File: src/stdlib/math.rs (hunk 3 above)
Comment (19wintersp) [reply src/stdlib/math.rs:46]: If that is the case, I'd ask that you do check carefully as to how the overall system works, at least for the general types and errors. (CC @cursorweb: we should really have some developer documentation)

File: src/types/tostring.rs (hunk 2 above)
Comment (AntimatterReactor) [reply src/types/tostring.rs:38]: Fair enough then.
Will do that if possible.

File: src/stdlib/math.rs (hunk 3 above)
Comment (AntimatterReactor) [reply src/stdlib/math.rs:46]: ok

File: Cargo.toml (hunk 1 above)
Comment (AntimatterReactor) [reply Cargo.toml:13]: Done!

File: src/stdlib/math.rs (hunk 4 above)
Comment (AntimatterReactor) [reply src/stdlib/math.rs:40]: After further reading, I don't think that replacing gen_range is possible with rand::random?

Comment (19wintersp) [reply src/stdlib/math.rs:40]: As I recommended, that would be the best option if you're not able to instantiate a single instance of `thread_rng` at initialisation.

File: src/functions/traits.rs [hunk 5]
```diff
@@ -1,6 +1,6 @@
-use crate::{error::Error, interpreter::Interpreter, types::Type};
+use crate::{error::ErrorType, interpreter::Interpreter, types::Type};
 
-pub type FResult = Result<Type, Error>;
+pub type FResult = Result<Type, (String, ErrorType)>;
```
Comment (19wintersp) [inline src/functions/traits.rs:3]: Why was this changed?

File: src/stdlib/math.rs (hunk 3 above)
Comment (AntimatterReactor) [reply src/stdlib/math.rs]: Solved! Kinda.

File: src/interpreter.rs [hunk 6]
```diff
@@ -276,7 +276,10 @@ impl Interpreter {
                         ));
                     }
 
-                    return func.call(self, params);
+                    return match func.call(self, params){
+                        Ok(v) => Ok(v),
+                        Err(e) => Err(Error::new(tok.lineinfo, e.0, e.1)),
+                    };
```
Comment (19wintersp) [inline src/interpreter.rs:282]: I don't think this should be changed. This is a hack - we need proper tracebacks on errors by this point.

File: src/stdlib/math.rs [hunk 7]
```diff
@@ -23,26 +24,43 @@ pub fn new() -> Module {
...
-                        .expect("Invalid Argument: Type is not Int")
-                ) as f32))
+                if let (Ok(min), Ok(max)) = (args[0].parse::<f32>(), args[1].parse::<f32>()) {
+                    Ok(Type::Float(thread_rng().gen_range(min..max)))
+                } else {
+                    Err((
+                        "Incorrect Argument Type: Type is not Int".into(),
+                        ErrorType::TypeError,
+                    ))
+                }
```
Comment (19wintersp) [inline src/stdlib/math.rs:63]: I asked that you change this: please stop using the `parse` method of the `Type`, and start actually checking the type and destructuring the argument.

**Review by 19wintersp — CHANGES_REQUESTED** [review]
Please address the issues from my previous review.

Why was the type of `FResult` changed?

File: src/stdlib/math.rs (hunk 3 above)
Comment (19wintersp) [reply src/stdlib/math.rs]: This was not resolved.

File: src/functions/traits.rs (hunk 5 above)
Comment (AntimatterReactor) [reply src/functions/traits.rs:3]: I just changed it because it's faster to do that than to replace FResult one-by-one to do the traceback hack.

Comment (19wintersp) [reply src/functions/traits.rs:3]: We shouldn't be using hacks. If there is a problem, it needs to be addressed (probably in a different PR) instead of being worked around. The fact that you've changed this means now all of the other functions have to deal with it. A better solution would be to fix the specific problem at hand: to make the line information *optional*. **Do not try to replace the** `Error` **type.**

File: src/interpreter.rs (hunk 6 above)
Comment (AntimatterReactor) [reply src/interpreter.rs:282]: I know, I just add it there so that it actually gives off the correct lineinfo value. Which is it's a temporary fix

Comment (19wintersp) [reply src/interpreter.rs:282]: We don't want temporary fixes, this code is going to be effectively permanent. If tracebacks are required, they should probably be implemented now before this is merged. Another issue is that this means errors inside user functions will not have their true source reported.

File: src/stdlib/math.rs (hunk 7 above)
Comment (AntimatterReactor) [reply src/stdlib/math.rs]: Done, Happy Now?

Comment (19wintersp) [reply src/stdlib/math.rs]: This is even worse; I've already said twice to stop trying to do fancy things, and just destructure it!

This is not resolved.

**Review by 19wintersp — CHANGES_REQUESTED** [review]
My previous comments still stand, please don't mark them as resolved yet.

File: src/stdlib/math.rs (hunk 7 above)
Comment (19wintersp) [reply src/stdlib/math.rs]: Your code means that `cos("1.23")` is valid - it shouldn't be.

**Comment by AntimatterReactor** [comment]
You know what? This PR has gone way off base. I definitely don't want to also fix the way function traceback are implemented. Closing...
